src/structures/vroom/amount.h

```diff
@@ -14,32 +14,151 @@ All rights reserved (see LICENSE).
 
 #include "structures/typedefs.h"
 
-class amount_t : private std::vector<capacity_t> {
+template <typename E> class amount_expression_t {
+public:
+  capacity_t operator[](size_t i) const {
+    return static_cast<const E&>(*this)[i];
+  };
+  std::size_t size() const {
+    return static_cast<const E&>(*this).size();
+  };
+  bool empty() const {
+    return size() == 0;
+  };
+};
 
-  using parent = std::vector<capacity_t>;
+template <typename E1, typename E2>
+bool operator<(const amount_expression_t<E1>& lhs,
+               const amount_expression_t<E2>& rhs) {
+  bool is_strict_inf = true;
+  assert(lhs.size() == rhs.size());
+  for (std::size_t i = 0; i < lhs.size(); ++i) {
+    if (lhs[i] >= rhs[i]) {
+      is_strict_inf = false;
+      break;
+    }
+  }
+
+  return is_strict_inf;
+}
+
+template <typename E1, typename E2>
+bool operator<=(const amount_expression_t<E1>& lhs,
+                const amount_expression_t<E2>& rhs) {
+  bool is_inf = true;
+  assert(lhs.size() == rhs.size());
+  for (std::size_t i = 0; i < lhs.size(); ++i) {
+    if (lhs[i] > rhs[i]) {
+      is_inf = false;
+      break;
+    }
+  }
+
+  return is_inf;
+}
+
+template <typename E1, typename E2>
+bool operator==(const amount_expression_t<E1>& lhs,
+                const amount_expression_t<E2>& rhs) {
+  bool is_equal = true;
+  assert(lhs.size() == rhs.size());
+  for (std::size_t i = 0; i < lhs.size(); ++i) {
+    if (lhs[i] != rhs[i]) {
+      is_equal = false;
+      break;
+    }
+  }
+
+  return is_equal;
+}
+
+class amount_t : public amount_expression_t<amount_t> {
+
+  std::vector<capacity_t> elems;
 
 public:
-  amount_t();
-  amount_t(std::size_t size);
+  amount_t() = default;
+
+  amount_t(std::size_t size) : elems(size, 0){};
+
+  template <typename E> amount_t(const amount_expression_t<E>& u) {
+    elems.resize(u.size());
+    for (std::size_t i = 0; i < u.size(); ++i) {
+      elems[i] = u[i];
+    }
+  }
+
+  void push_back(capacity_t c) {
+    elems.push_back(c);
+  };
+
+  capacity_t operator[](std::size_t i) const {
+    return elems[i];
+  };
+
+  capacity_t& operator[](std::size_t i) {
+    return elems[i];
+  };
 
-  using parent::empty;
-  using parent::size;
-  using parent::operator[];
-  using parent::push_back;
+  std::size_t size() const {
+    return elems.size();
+  };
 
   amount_t& operator+=(const amount_t& rhs);
```

Comment:
I guess the reason for having all implementations for expression templates in the `.h` file is the same as for my question in the previous PR (allows the compiler to have the details from other compilation units)?

If so, wouldn't that make sense for `amount_t::operator+=` and `amount_t::operator-=` too?

Reply:
I moved them to the header and deleted the .cpp file, since it is now empty.